src/soundvectors.py

```diff
@@ -1,353 +1,439 @@
+    rising: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    falling: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    contour: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    backshift: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    frontshift: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    opening: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    closing: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    centering: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    longdistance: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+    secondrounded: int = dataclasses.field(default=0, metadata={'dc:description': ''})
+
+    def __post_init__(self):
+        assert all(v in {0, 1, -1} for v in dataclasses.astuple(self))
+
+    @classmethod
+    def from_positive_and_negative(cls, pos: list, neg: list):
+        """
+        Factory method to allow for transparent instantiation.
+        """
+        assert not set(pos).intersection(neg)
+        return cls(**{k: 1 if k in pos else -1 for k in pos + neg})
+
+    def as_vector(self) -> typing.Tuple[int]:
+        return dataclasses.astuple(self)
 
-binary_features = ["cons", "syl", "son", "cont", "delrel", "lat", "nas", "voi", "sg", "cg", "pharyngeal", "laryngeal",
-    "cor", "dorsal", "lab", "hi", "lo", "back", "front", "tense", "round", "velaric", "long", "ant", "distr", "strid",
-    "hitone", "hireg", "loreg", "rising", "falling", "contour", "backshift", "frontshift", "opening", "closing",
-    "centering", "longdistance", "secondrounded", ]
+    def as_dict(self,
+                valid_values: typing.Optional[typing.Union[typing.Container, typing.Iterable]]
+                = None) -> typing.Dict[str, int]:
+        """
+        Returns a `dict` keyed with the fields of the FeatureBundle, filtered to values in
+        `valid_values`.
+        """
+        return {
+            k: v for k, v in dataclasses.asdict(self).items()
+            if valid_values is None or (v in valid_values)}
+
+    def as_set(self):
+        return frozenset(
+            {'{}{}'.format('+' if v > 0 else '-', k) for k, v in self.as_dict({1, -1}).items()})
+
+    def __str__(self):
+        return ','.join(
+            '{}{}'.format('+' if v > 0 else ('-' if v < 0 else '0_'), k)
+            for k, v in self.as_dict().items())
 
-clts_feature_hierarchy = {"type": 0, "manner": 1, "height": 1, "roundedness": 1, "centrality": 1, "place": 2,
+    def replace(self, **changes):
+        """
+        Shortcut for `dataclasses.replace`.
+        """
+        return dataclasses.replace(self, **changes)
+
+    @classmethod
+    def fields(cls):
+        """
+        Shortcut for `dataclasses.fields`.
+        """
+        return dataclasses.fields(cls)
```

Comment:
A `FeatureBundle` proxies the `dataclasses` functions `fields` and `replace` such that code that uses it doesn't need to import `dataclasses`.